### PR TITLE
generator refactoring

### LINKED ISSUE #1323: <experimental/generator>: error C2980 when compiling with /kernel (regression)
**Describe the bug**
Following the changes from PR #1219, using the `experimental/generator` header in code built with `/kernel` (and hence no exception handling) now fails to compile with error C2980.

**Command-line test case**
```
C:\Users\kobyk\Documents\Stuff\ge>type ge.cpp
#include <malloc.h>
#include <coroutine>
#include <experimental/generator>

struct WDFQUEUE{};

namespace foo {

template <typename T>
class PoolAllocator final {
public:
    using value_type = T;
    using pointer = T*;
    using const_pointer = const T*;
    using reference = T&;
    using const_reference = const T&;
    using size_type = std::size_t;
    using difference_type = std::ptrdiff_t;

    template <class U>
    struct rebind {
        using other = PoolAllocator<T>;
    };

    pointer allocate(size_t size) noexcept {
        return static_cast<pointer>(malloc(size));
    }

    void deallocate(pointer ptr, [[maybe_unused]] size_t size) noexcept {

        free(ptr);
    }
};

template <typename T>
using Generator = std::experimental::generator<T, PoolAllocator<char>>;

class IteratedRequest final
{
public:
        bool valid() const { return true; }
        bool next(WDFQUEUE queue) { return true; }
};

Generator<IteratedRequest> IterateRequests(WDFQUEUE queue) {

    IteratedRequest currentRequest;
    while (currentRequest.next(queue)) {
        if (!currentRequest.valid()) {
            continue;
        }
        co_yield currentRequest;
    }
}

}
C:\Users\kobyk\Documents\Stuff\ge>cl /kernel /nologo /std:c++latest /c ge.cpp
ge.cpp
ge.cpp(54): error C2980: C++ exception handling is not supported with /kernel
```

**Expected behavior**
Previously, generators with a non-throwing allocator worked with code built with `/kernel`.
Furthermore, removing the `unhandled_exception()` function added by the PR results in the code compiling successfully.

**STL version**
Microsoft Visual Studio Community 2019 Preview
Version 16.8.0 Preview 3.1

**Additional context**
Switching back to legacy coroutines (i.e., including `<experimental/coroutine>` instead of `<coroutine>` and adding `/await` to the compiler command line) also resolves the C2980 error.

@CaseyCarter

### PR DESCRIPTION
**Pull request by CaseyCarter**
Restore kernel mode operation by defining `unhandled_exception` only when `_KERNEL_MODE` is not defined. (When `unhandled_exception` is defined, the compiler will wrap the coroutine body in a `try...catch` as depicted in [dcl.fct.def.coroutine]/5, which it then diagnoses as ill-formed when unwinding is disabled as in `/kernel`.)

Fixes #1323.

Some cleanup while we're here:
* Use STL style:
  * Prefer not to deduce simple return types
  * `const` west
  * introduce template type parameters with `class`
  * Annotate functions that don't throw exceptions `noexcept`
  * remove excess empy lines
  * Use `_Capital_snake` names for non-template-parameters
  * `generator::promise_type` only has members `_Exception` and `_Rethrow_if_exception` to "delay" exception delivery so as to avoid triggering VSO-1172852. They aren't needed in kernel mode.
* `generator::iterator` should have _explicit_ conversion from `coroutine_handle` rather than implicit for good design hygiene. The conversion from `nullptr_t` is unnecessary.

Note that the story for `/EHs-` without `/kernel` is less clear and currently not supported.

### CODE REVIEW

File: stl/inc/experimental/generator [hunk 1]
```diff
@@ -33,18 +32,18 @@ _STL_DISABLE_CLANG_WARNINGS
 _STD_BEGIN
 
 namespace experimental {
-
-    template <typename _Ty, typename _Alloc = allocator<char>>
+    // NOTE WELL: _CPPUNWIND currently affects the ABI of generator. Linking together /EHs
+    // translation units and /EHs- translation units that use the same specialization of generator
+    // violates the One Definition Rule. This could result in fantastic failures or obscure bugs.
```
Comment (CaseyCarter) [inline stl/inc/experimental/generator:37]: This comment should avoid `/EHs-` which isn't supported:
```suggestion
    // NOTE WELL: _CPPUNWIND currently affects the ABI of generator.
```
(I don't think we need to mention linking /kernel and non-/kernel TUs which would be a bizarre thing to do regardless.)

**Review by GorNishanov — APPROVED** [review]
LGTM!

**Review by joemmett — APPROVED** [review]

File: stl/inc/experimental/generator [hunk 2]
```diff
@@ -56,38 +55,38 @@ namespace experimental {
...
+                _Exception = _STD current_exception();
+            }
 #else // ^^^ workaround / no workaround vvv
+            void unhandled_exception() {
                 throw;
-#endif // TRANSITION, VSO-1172852
             }
-#endif // defined(_CPPUNWIND) || defined(__cpp_impl_coroutine)
+#endif // TRANSITION, VSO-1172852
+#else // ^^^ defined(_CPPUNWIND / !defined(_CPPUNWIND)
```
Comment (cbezault) [inline stl/inc/experimental/generator:69]:
```suggestion
#else // ^^^ defined(_CPPUNWIND / !defined(_CPPUNWIND) vvv

```

**Review by cbezault — APPROVED** [review]

File: stl/inc/experimental/generator [hunk 3]
```diff
@@ -33,18 +32,16 @@ _STL_DISABLE_CLANG_WARNINGS
 _STD_BEGIN
 
 namespace experimental {
-
-    template <typename _Ty, typename _Alloc = allocator<char>>
+    // NOTE WELL: _CPPUNWIND currently affects the ABI of generator.
```
Comment (StephanTLavavej) [inline stl/inc/experimental/generator:35]: Should we `#pragma detect_mismatch` in this header? (OK if the answer is "no" or "we can file an enhancement issue".)

File: stl/inc/experimental/generator [hunk 4]
```diff
@@ -115,21 +113,19 @@ namespace experimental {
...
+            using reference         = const _Ty&;
+            using pointer           = const _Ty*;
 
             coroutine_handle<promise_type> _Coro = nullptr;
 
             iterator() = default;
-            iterator(nullptr_t) : _Coro(nullptr) {}
-
-            iterator(coroutine_handle<promise_type> _CoroArg) : _Coro(_CoroArg) {}
+            explicit iterator(coroutine_handle<promise_type> _CoroArg) noexcept : _Coro(_CoroArg) {}
```
Comment (StephanTLavavej) [inline stl/inc/experimental/generator:122]: Pre-existing naming ultra-nitpick, `_CoroArg` is `_UglyCamel` but we generally use `_Ugly_snake` for variables. No change requested.

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/experimental/generator (hunk 3 above)
Comment (CaseyCarter) [reply stl/inc/experimental/generator:35]: It feels like overkill to guard against mixing kernel and non-kernel TUs with detect_mismatch. It made a lot of sense to have a comment here when I was expecting to support `/EHs-` with the `/kernel` layout - linking `/EHs` and `/EHs-` TUs seems more likely than `/kernel` and non-`/kernel` - but now I'm not sure the comment adds value at all.

File: stl/inc/experimental/generator (hunk 4 above)
Comment (CaseyCarter) [reply stl/inc/experimental/generator]: Thanks - that's exactly the kind of unconventional usage I'm trying to correct here.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this regression! I was unable to think of a clever pun involving `/kernel` and :popcorn: here, so you'll have to experiment with generating one yourself. :joy_cat:

**Comment by CaseyCarter** [comment]
It was Colonel 🍿, in the conservatory, with the lead pipe!

**Pull request merged**